src/Adapter/Product/CombinationDataProvider.php

```diff
@@ -0,0 +1,178 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Product;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Query\QueryBuilder;
+use PrestaShop\PrestaShop\Core\Domain\Language\ValueObject\LanguageId;
+
+/**
+ * Provides data associated with product combinations
+ */
+class CombinationDataProvider
+{
+    /**
+     * @var Connection
+     */
+    private $connection;
+
+    /**
+     * @var string
+     */
+    private $dbPrefix;
+
+    /**
+     * @param Connection $connection
+     * @param string $dbPrefix
+     */
+    public function __construct(
+        Connection $connection,
+        string $dbPrefix
+    ) {
+        $this->connection = $connection;
+        $this->dbPrefix = $dbPrefix;
+    }
+
+    /**
+     * @param int $productId
+     * @param int $limit
+     * @param int $offset
+     * @param array $filters
+     *
+     * @return array
+     */
+    public function getCombinationsInfo(int $productId, int $limit, int $offset = 0, array $filters = []): array
+    {
+        //@todo: filters not handled.
+        $qb = $this->getCombinationsQueryBuilder($productId)
+            ->select('pa.*')
+            ->setParameter('productId', $productId)
+            ->setFirstResult($offset)
+            ->setMaxResults($limit)
+        ;
+
+        return [
+            'combinations' => $qb->execute()->fetchAll(),
+            'total_count' => $this->getTotalCombinationsCount($productId),
+        ];
+    }
+
+    /**
+     * @param int[] $combinationIds
+     * @param LanguageId $langId
+     *
+     * @return array
+     */
+    public function getAttributesInfoByCombinationIds(array $combinationIds, LanguageId $langId): array
+    {
+        $qb = $this->connection->createQueryBuilder();
```

Comment:
maybe we could separate these query calls into private functions, this would make it easier to read and understand: `getAssociatedAttributes`, `getAttributesInfos`, ...